In MS Scenarios - No RewardClaimed event found for this channel

## Problem Definition 

In [MS1](https://github.com/raiden-network/raiden/blob/develop/raiden/tests/scenarios/ms1_simple_monitoring.yaml), [MS2](https://github.com/raiden-network/raiden/blob/develop/raiden/tests/scenarios/ms2_simple_monitoring.yaml) and [BF1](https://github.com/raiden-network/raiden/blob/develop/raiden/tests/scenarios/bf1_basic_functionality.yaml) we want to see the Raiden monitoring service to react properly. Monitoring works and the Monitoring Service reacts as expected. 

However, those scenarios fail constantly when the Monitoring Service tries to claim its reward with 

```
ScenarioAssertionError(\"No RewardClaimed event found for this channel.\") scenario_player.exceptions.legacy.ScenarioAssertionError: No RewardClaimed event found for this channel.",
```

The calculation of the point in time at which this reward is to be claimed seems wrong.

--- a/src/monitoring_service/handlers.py
+++ b/src/monitoring_service/handlers.py
@@ -329,11 +329,16 @@ def monitor_new_balance_proof_event_handler(event: Event, context: Context) -> N
 
         context.database.upsert_channel(channel)
 
-    # check if this was our update, if so schedule the call
-    # of `claimReward`
-    # it will be checked there that our update was the latest one
+    # check if this was our update, if so schedule the call of
+    # `claimReward` it will be checked there that our update was the latest one
     if event.ms_address == context.ms_state.address:
         assert channel.closing_block is not None, "closing_block not set"
+        # Transactions go into the next mined block, so we could trigger one block
+        # before the `claim` call is allowed to succeed to include it in the
+        # first possible block.
+        # Unfortunately, parity does the gas estimation on the current block
+        # instead of the next one, so we have to wait for the first allowed
+        # block to be finished to send the transaction successfully on parity.
         trigger_block = BlockNumber(channel.closing_block + channel.settle_timeout + 1)
 
         # trigger the claim reward action by an event

--- a/src/monitoring_service/handlers.py
+++ b/src/monitoring_service/handlers.py
@@ -334,20 +334,30 @@ def monitor_new_balance_proof_event_handler(event: Event, context: Context) -> N
     # it will be checked there that our update was the latest one
     if event.ms_address == context.ms_state.address:
         assert channel.closing_block is not None, "closing_block not set"
-        trigger_block = BlockNumber(channel.closing_block + channel.settle_timeout)
+        trigger_block = BlockNumber(channel.closing_block + channel.settle_timeout + 1)
 
         # trigger the claim reward action by an event
-        event = ActionClaimRewardTriggeredEvent(
+        triggered_event = ActionClaimRewardTriggeredEvent(
             token_network_address=channel.token_network_address,
             channel_identifier=channel.identifier,
             non_closing_participant=event.raiden_node_address,
+        )
+
+        log.info(
+            "Received event for own update, triggering reward claim",
+            token_network_address=event.token_network_address,
+            identifier=channel.identifier,
+            scheduled_event=triggered_event,
+            trigger_block=trigger_block,
+            closing_block=channel.closing_block,
+            settle_timeout=channel.settle_timeout,
         )
 
         # Add scheduled event if it not exists yet
         # If the event is already scheduled (e.g. after a restart) the DB takes care that
         # it is only stored once
         context.database.upsert_scheduled_event(
-            ScheduledEvent(trigger_block_number=trigger_block, event=cast(Event, event))
+            ScheduledEvent(trigger_block_number=trigger_block, event=cast(Event, triggered_event))
         )
 
 

--- a/src/monitoring_service/service.py
+++ b/src/monitoring_service/service.py
@@ -58,7 +58,12 @@ def handle_event(event: Event, context: Context) -> None:
     Exceptions are caught and generate both error logs and sentry issues.
     Events are not retried after an exception.
     """
-    log.debug("Processing event", event_=event)
+    log.debug(
+        "Processing event",
+        event_=event,
+        latest_confirmed_block=context.latest_confirmed_block,
+        latest_unconfirmed_block=context.get_latest_unconfirmed_block(),
+    )
     handler = HANDLERS.get(type(event))
 
     if handler:

--- a/tests/monitoring/monitoring_service/test_end_to_end.py
+++ b/tests/monitoring/monitoring_service/test_end_to_end.py
@@ -1,5 +1,7 @@
+from typing import Callable
+
 import gevent
-from eth_utils import decode_hex, encode_hex, to_canonical_address, to_checksum_address
+from eth_utils import decode_hex, encode_hex, to_canonical_address
 from request_collector.server import RequestCollector
 from web3 import Web3
 
@@ -24,7 +26,7 @@ from raiden_libs.blockchain import query_blockchain_events
 
 def create_ms_contract_events_query(
     web3: Web3, contract_manager: ContractManager, contract_address: Address
-):
+) -> Callable:
     def f():
         return query_blockchain_events(
             web3=web3,
@@ -57,14 +59,13 @@ def test_first_allowed_monitoring(
     query = create_ms_contract_events_query(
         web3, contracts_manager, monitoring_service_contract.address
     )
-    ms_address_hex = to_checksum_address(monitoring_service.address)
     c1, c2 = get_accounts(2)
 
     # add deposit for c1
     node_deposit = 10
     deposit_to_udc(c1, node_deposit)
 
-    assert service_registry.functions.hasValidRegistration(ms_address_hex).call()
+    assert service_registry.functions.hasValidRegistration(monitoring_service.address).call()
 
     # each client does a transfer
     channel_id = create_channel(c1, c2, settle_timeout=10)[0]
@@ -176,15 +177,14 @@ def test_e2e(  # pylint: disable=too-many-arguments,too-many-locals
     query = create_ms_contract_events_query(
         web3, contracts_manager, monitoring_service_contract.address
     )
-    ms_address_hex = to_checksum_address(monitoring_service.address)
-    initial_balance = user_deposit_contract.functions.balances(ms_address_hex).call()
+    initial_balance = user_deposit_contract.functions.balances(monitoring_service.address).call()
     c1, c2 = get_accounts(2)
 
     # add deposit for c1
     node_deposit = 10
     deposit_to_udc(c1, node_deposit)
 
-    assert service_registry.functions.hasValidRegistration(ms_address_hex).call()
+    assert service_registry.functions.hasValidRegistration(monitoring_service.address).call()
 
     # each client does a transfer
     channel_id = create_channel(c1, c2, settle_timeout=5)[0]
@@ -242,28 +242,17 @@ def test_e2e(  # pylint: disable=too-many-arguments,too-many-locals
     ).transact({"from": c2})
     # Wait until the MS reacts, which it does after giving the client some time
     # to update the channel itself.
+
     wait_for_blocks(2)  # 1 block for close + 1 block for triggering the event
     # Now give the monitoring service a chance to submit the missing BP
     gevent.sleep(0.01)
-
     assert [e.event for e in query()] == [MonitoringServiceEvent.NEW_BALANCE_PROOF_RECEIVED]
 
     # wait for settle timeout
-    wait_for_blocks(2)  # timeout is 5, but we've already waited 3 blocks before
+    # timeout is 5, but we've already waited 3 blocks before. Additionally one block is
+    # added to handle parity running gas estimation on current instead of next.
+    wait_for_blocks(3)
 
-    token_network.functions.settleChannel(
-        channel_id,
-        c1,  # participant_B
-        transferred_c1,  # participant_B_transferred_amount
-        0,  # participant_B_locked_amount
-        LOCKSROOT_OF_NO_LOCKS,  # participant_B_locksroot
-        c2,  # participant_A
-        transferred_c2,  # participant_A_transferred_amount
-        0,  # participant_A_locked_amount
-        LOCKSROOT_OF_NO_LOCKS,  # participant_A_locksroot
-    ).transact()
-
-    # Wait until the ChannelSettled is confirmed
     # Let the MS claim its reward
     gevent.sleep(0.01)
     assert [e.event for e in query()] == [
@@ -271,7 +260,7 @@ def test_e2e(  # pylint: disable=too-many-arguments,too-many-locals
         MonitoringServiceEvent.REWARD_CLAIMED,
     ]
 
-    final_balance = user_deposit_contract.functions.balances(ms_address_hex).call()
+    final_balance = user_deposit_contract.functions.balances(monitoring_service.address).call()
     assert final_balance == (initial_balance + reward_amount)
 
     ms_greenlet.kill()

--- a/tests/monitoring/monitoring_service/test_handlers.py
+++ b/tests/monitoring/monitoring_service/test_handlers.py
@@ -1,4 +1,5 @@
 # pylint: disable=redefined-outer-name
+from typing import Optional
 from unittest.mock import Mock, patch
 
 import pytest
@@ -17,6 +18,7 @@ from monitoring_service.database import Database
 from monitoring_service.events import (
     ActionClaimRewardTriggeredEvent,
     ActionMonitoringTriggeredEvent,
+    ScheduledEvent,
 )
 from monitoring_service.handlers import (
     Context,
@@ -100,6 +102,20 @@ def setup_state_with_closed_channel(context: Context) -> Context:
     assert_channel_state(context, ChannelState.CLOSED)
 
     return context
+
+
+def get_scheduled_claim_event(database: Database) -> Optional[ScheduledEvent]:
+    events = database.get_scheduled_events(max_trigger_block=BlockNumber(999_999))
+
+    filtered_events = [
+        event for event in events if isinstance(event.event, ActionClaimRewardTriggeredEvent)
+    ]
+    assert len(filtered_events) <= 1
+
+    if len(filtered_events) == 0:
+        return None
+
+    return filtered_events[0]
 
 
 @pytest.fixture
@@ -345,7 +361,7 @@ def test_monitor_new_balance_proof_event_handler_sets_update_status(context: Con
         nonce=Nonce(2),
         ms_address=Address(bytes([4] * 20)),
         raiden_node_address=DEFAULT_PARTICIPANT2,
-        block_number=BlockNumber(23),
+        block_number=BlockNumber(62),
     )
 
     channel = context.database.get_channel(
@@ -353,6 +369,7 @@ def test_monitor_new_balance_proof_event_handler_sets_update_status(context: Con
     )
     assert channel
     assert channel.update_status is None
+    assert get_scheduled_claim_event(context.database) is None
 
     monitor_new_balance_proof_event_handler(new_balance_event, context)
 
@@ -365,6 +382,10 @@ def test_monitor_new_balance_proof_event_handler_sets_update_status(context: Con
     assert channel.update_status.nonce == 2
     assert channel.update_status.update_sender_address == bytes([4] * 20)
 
+    scheduled_claim_event = get_scheduled_claim_event(context.database)
+    assert scheduled_claim_event is not None
+    assert scheduled_claim_event.trigger_block_number == 52 + 100 + 1
+
     new_balance_event2 = ReceiveMonitoringNewBalanceProofEvent(
         token_network_address=DEFAULT_TOKEN_NETWORK_ADDRESS,
         channel_identifier=DEFAULT_CHANNEL_IDENTIFIER,
@@ -372,7 +393,7 @@ def test_monitor_new_balance_proof_event_handler_sets_update_status(context: Con
         nonce=Nonce(5),
         ms_address=Address(bytes([4] * 20)),
         raiden_node_address=DEFAULT_PARTICIPANT2,
-        block_number=BlockNumber(23),
+        block_number=BlockNumber(63),
     )
 
     monitor_new_balance_proof_event_handler(new_balance_event2, context)
@@ -385,6 +406,10 @@ def test_monitor_new_balance_proof_event_handler_sets_update_status(context: Con
     assert channel.update_status is not None
     assert channel.update_status.nonce == 5
     assert channel.update_status.update_sender_address == bytes([4] * 20)
+
+    scheduled_claim_event = get_scheduled_claim_event(context.database)
+    assert scheduled_claim_event is not None
+    assert scheduled_claim_event.trigger_block_number == 52 + 100 + 1
 
 
 def test_monitor_new_balance_proof_event_handler_idempotency(context: Context):

